File: Java/028_implementStr.java
/*************************************************************************
    > File Name: 028_implementStr.java
    > Author: daiyu	
	> Url: https://leetcode.com/problems/implement-strstr/
    > Created Time: 五  9/18 15:34:44 2015
 ************************************************************************/
public class Solution {
    public int strStr(String haystack, String needle) {
        int result = 0;
        int nLength = needle.length();
        int hLength = haystack.length();
        //父字符串的长度要比子串长
        if (hLength < nLength) {
            return -1;
        }
        //子串长度为0        
        if (needle.length() == 0) {
            return 0;
        }
        //双重循环，利用m作为父字符串的工作指针，i纪录结果点，保持不动
        for (int i = 0; i < hLength; i ++) {
            if (i + nLength > hLength) {
                 return -1;
            }
            
            int m = i;
            for (int j = 0; j < nLength; j ++) {
                if(needle.charAt(j) == haystack.charAt(m)){
                    if(j == needle.length() - 1)
                        return i;
                    m++;
                } else {
                    break;
                }
            }
        }      
        
        return -1;
    }
}

File: Python/028_ImplementStr.py
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        result = 0
        h = len(haystack)
        n = len(needle)
        
        if h < n:
            return -1
        
        if n == 0:
            return 0
            
        for i in range(h):
            if i + n > h:
                #此时剩余子串字符串以不足作为一个子字符串，返回-1
                return -1
            
            tmp = i
            for j in range(n):
                if needle[j] == haystack[tmp]:
                    if j == n - 1:
                        return i
                    tmp += 1
                else:
                    break
                
        return -1